lib/Conversion/TorchToTosa/TorchToTosa.cpp

```diff
@@ -3702,28 +3702,67 @@ LogicalResult ConvertAtenOp<AtenArangeStartStepOp>::matchAndRewrite(
         op, "unimplemented: pin_memory must be either None or false");
   }
 
-  int64_t start, step, end;
-  if (!matchPattern(op.getStart(), m_TorchConstantInt(&start)))
+  auto matchIntOrDouble =
+      [&](Value val) -> std::tuple<LogicalResult, int64_t, double> {
+    // Match int or fp values. The one used depends on the resultType.
+    // Therefore `valueInt` and `valueDouble` will have similar values (but may
+    // be truncated due to casting).
+    int64_t valueInt = 0;
+    double valueDouble = 0.0;
+    if (matchPattern(val, m_TorchConstantInt(&valueInt)))
+      return {success(), valueInt, static_cast<double>(valueInt)};
+    if (matchPattern(val, m_TorchConstantFloat(&valueDouble)))
+      return {success(), static_cast<int64_t>(valueDouble), valueDouble};
+    return {failure(), valueInt, valueDouble};
+  };
+
+  auto [matchStart, startInt, startDouble] = matchIntOrDouble(op.getStart());
+  if (failed(matchStart))
     return rewriter.notifyMatchFailure(
-        op, "unimplemented: value `start` should be a torch constant int");
+        op,
+        "unimplemented: value `start` should be a torch constant int or float");
 
-  if (!matchPattern(op.getEnd(), m_TorchConstantInt(&end)))
+  auto [matchEnd, endInt, endDouble] = matchIntOrDouble(op.getEnd());
+  if (failed(matchEnd))
     return rewriter.notifyMatchFailure(
-        op, "unimplemented: value `end` should be a torch constant int");
+        op,
+        "unimplemented: value `end` should be a torch constant int or float");
 
-  if (!matchPattern(op.getStep(), m_TorchConstantInt(&step)))
+  auto [matchStep, stepInt, stepDouble] = matchIntOrDouble(op.getStep());
+  if (failed(matchStep))
     return rewriter.notifyMatchFailure(
-        op, "unimplemented: value `step` should be a torch constant int");
+        op,
+        "unimplemented: value `step` should be a torch constant int or float");
 
   // The result will always be a 1-d tensor.
   // The size of the result is calculated as follows:
   //          ceil((end - start)/step)
-  int64_t resultShape = ceil((float)(end - start) / (float)step);
-  SmallVector<int64_t> values(resultShape, start);
-  for (unsigned i = 1; i < resultShape; i++)
-    values[i] += i * step;
-  Value result =
-      tosa::getConstTensor<int64_t>(rewriter, op, values, resultShape).value();
+  auto elementType = resultType.getElementType();
+  Value result;
+  if (isa<mlir::IntegerType>(elementType)) {
+    int64_t resultShape = ceil(static_cast<double>(endInt - startInt) /
+                               static_cast<double>(stepInt));
+    SmallVector<int64_t> values(resultShape, startInt);
+    for (unsigned i = 1; i < resultShape; i++)
+      values[i] += i * stepInt;
+    result = tosa::getConstTensor<int64_t>(rewriter, op, values, resultShape)
```

Comment:
I don't quite see how this makes sure it has the right `elementType` - don't you need to at check `<intType>.getWidth()`?.

Reply:
We can either create a vector with the right type as we do for `int64_t`, `f64` or `f32`, or we construct the vector with a different type that has higher accuracy, and cast to the selected `elementType`. The `cast` will potentially be folded away I guess.